apis/common/v1/resource.go

```diff
@@ -161,55 +161,31 @@ type ResourceStatus struct {
 	ConditionedStatus `json:",inline"`
 }
 
-// A ProviderSpec defines the common way to get to the necessary objects to
-// connect to the provider.
-// Deprecated: Please use ProviderConfigSpec.
-type ProviderSpec struct {
-	// CredentialsSecretRef references a specific secret's key that contains
-	// the credentials that are used to connect to the provider.
-	// +optional
-	CredentialsSecretRef *SecretKeySelector `json:"credentialsSecretRef,omitempty"`
+// CredentialsSelectors provides common selectors for extracting credentials.
+type CredentialsSelectors struct {
+	// Fs is a reference to a filesystem location that contains credentials that
+	// must be used to connect to the provider. +optional
+	Fs *FsSelector `json:"fs,omitempty"`
+
+	// Env is a reference to an environment variable that contains credentials
+	// that must be used to connect to the provider. +optional
+	Env *EnvSelector `json:"env,omitempty"`
+
+	// A SecretRef is a reference to a secret key that contains the credentials
+	// that must be used to connect to the provider. +optional
+	SecretRef *SecretKeySelector `json:"secretRef,omitempty"`
 }
 
-// A ProviderConfigSpec defines the desired state of a provider config. A
-// provider config may embed this type in its spec in order to support standard
-// fields. Provider configs may choose to avoid embedding this type as
-// appropriate, but are encouraged to follow its conventions.
-type ProviderConfigSpec struct {
-	// Credentials required to authenticate to this provider.
-	Credentials ProviderCredentials `json:"credentials"`
+// EnvSelector selects an environment variable.
+type EnvSelector struct {
+	// Name is the name of an environment variable.
+	Name string `json:"name,omitempty"`
 }
 
-// A CredentialsSource is a source from which provider credentials may be
-// acquired.
-type CredentialsSource string
```

Comment:
Per our discussion yesterday, I still prefer to keep this as an enum for UX purposes. Perhaps a good compromise would be to:

1. Keep this type definition here, along with a few of the most common credential sources that providers may (optionally) choose to support.
2. Remove the generic `ProviderConfigSpec`, and instead have each provider implement their own.
3. Define standard `EnvSelector`, `FsSelector`, `SecretKeySelector` etc structs here that providers may also choose to support.
4. Define simple helper functions for each type - i.e. a type that takes a `SecretKeySelector` and a client and returns the secret data. These functions wouldn't be attempting to satisfy a consistent type or interface - providers would be responsible for switching on the CredentialsSource and calling the appropriate function with the appropriate inputs (i.e. the appropriate selector struct).

Reply:
@negz I think we can still have the advantages the enum provides with the approach I have here by letting each provider define their `Source` as an enum of the types they support, right? This makes it such that we can still provide the default handlers and avoid forcing provider authors to implement their own switch. More info here https://github.com/crossplane/crossplane-runtime/issues/236#issuecomment-769062449

Reply:
@negz I opened a PR to show how this would look on a ProviderConfig CRD in practice. I think my extractors are going to have to be modified at least partially because we are going to have to pass a different struct for those defined as defaults vs additional selectors added by provider implementation, but the source enum works pretty well IMO. Let me know your thoughts https://github.com/crossplane/provider-aws/pull/518/files#diff-1201d66f172379e6b70b7ffed32ee04906876a8c748c6397e278ee3bf933e28aR107

Reply:
Thanks - that illustration helps a lot. This approach LGTM, though I'd suggest we define the `CredentialsSource` type and the common sources (none, secret, env, fs) in this package, rather than defining them in each provider given that AFAICT all providers will be expected to know about those sources (because they embed CredentialsSelectors). Each provider could still define and use additional sources.